go/oasis-test-runner/log/handlers.go

```diff
@@ -14,53 +23,94 @@ func (a *assertBase) fail() error {
 	return fmt.Errorf("log assertion failed: %s", a.message)
 }
 
-func (a *assertBase) Line(line string) error {
+func (a *assertBase) String() string {
+	return fmt.Sprintf("assertBase{message: %s}", a.message)
+}
+
+type assertBaseHandler assertBase
+
+func (h *assertBaseHandler) Line(line string) error {
 	return nil
 }
 
-func (a *assertBase) Finish() error {
+func (h *assertBaseHandler) Finish() error {
 	return nil
 }
 
+type assertBaseFactory assertBase
+
+func (fac *assertBaseFactory) Generate() (WatcherHandler, error) {
+	return &assertBaseHandler{
+		message: fac.message,
+	}, nil
+}
+
 type assertContains struct {
 	assertBase
 
 	text     string
-	seen     bool
 	expected bool
 }
 
-func (a *assertContains) Line(line string) error {
-	if strings.Contains(line, a.text) {
-		a.seen = true
+func (a *assertContains) String() string {
+	return fmt.Sprintf("assertContains{message: %s text: %s expected: %t}", a.message, a.text, a.expected)
+}
+
+type assertContainsHandler struct {
+	assertContains
+
+	seen bool
+}
+
+func (h *assertContainsHandler) Line(line string) error {
+	if strings.Contains(line, h.text) {
+		h.seen = true
 	}
 	return nil
 }
 
-func (a *assertContains) Finish() error {
-	if a.seen != a.expected {
-		return a.fail()
+func (h *assertContainsHandler) Finish() error {
+	if h.seen != h.expected {
+		return h.fail()
 	}
 	return nil
 }
 
-// AssertContains returns a handler which checks that given text is
-// contained in the log output.
-func AssertContains(text, message string) WatcherHandler {
-	return &assertContains{
-		assertBase: assertBase{message},
-		text:       text,
-		expected:   true,
+type assertContainsFactory struct {
+	assertContains
+}
+
+func (fac *assertContainsFactory) Generate() (WatcherHandler, error) {
+	return &assertContainsHandler{
+		assertContains: assertContains{
```

Comment:
Just passing `fac.assertContains` should be enough to create a copy (as composition is done by value not by pointer so value assignment should copy), so something like:
```suggestion
		assertContains: fac.assertContains,
```
Also elsewhere to simplify code.

Reply:
Thanks for the tip!

Done.